### DISCUSSION TITLE
[Question] Single Kiali for Monitoring Multi-cluster Istio Installation

### DISCUSSION THREAD

**Discussion by ethernoy** [kiali/kiali#3963] [Multicluster]
Hi all, I wonder if Kiali supports monitoring a multi-cluster Istio installation as illustrated in the following diagram?

![image](https://user-images.githubusercontent.com/15870761/116536749-e5d1b580-a917-11eb-9a76-421b6b509130.png)

Note that Cluster A and Cluster B are configured as documented in https://istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/. A B and Management clusters both have independent Istiod instances.

Assuming that Kiali can not directly access pod within A and B, but we can expose the Prometheus within A and B or let them remote write metric to Management Cluster.

**Comment by lucasponce** [comment]
Thanks for the question @ethernoy.

Deployment models in Kiali is a quite active topic, so I moved the issue into a discussion with a specific category for it.

**Reply by ethernoy** [reply]
> Thanks for the question @ethernoy.
> 
> Deployment models in Kiali is a quite active topic, so I moved the issue into a discussion with a specific category for it.

Thank you, I didn't realize this place is a better place for raising my question :)

**Reply by israel-hdez** [reply]
Hi @ethernoy 

At the moment, it is not possible and you will need to install Kiali on each cluster to monitor each portion of the Mesh.

I understand your Management cluster is isolated and on it's own mesh. If this is the case, Kiali will only let you monitor this mesh.

For cluster A and B, I see they are under "multi-primary on different networks". You still need to install Kiali on each cluster. Each Kiali will only let you see the portion of the mesh of the "local" cluster and some "border" services. However, the latest Kiali version added cross-linking in the graph to let you navigate between Kialis for a more seamless experience. This was demoed in our last demo which you can watch in youtube: https://youtu.be/qOvJ1YOWenU?t=1020

As Lucas said, lately this has been an active topic. I would like to get your thoughts/feedback on this topic.

I underestand that you have two meshes: one for the management cluster, and another one for cluster A and B. Do you really expect that Kiali shows a unified view of both meshes? Or do you expect Kiali to let you choose what mesh you want to inspect and then show *only* the selected mesh?

I'm asking because two independent meshes may also mean two different management domains, different users, different privileges, etc.

---

Now focusing only on *one* mesh: the one where Cluster A and B live:

I see that you would be comfortable exposing the Prometheus on each cluster. So, I guess you would expect that Kiali can be configured to talk to both Prometheuses.

There is also the alternative to configure a federated Prometheus ( see https://prometheus.io/docs/prometheus/latest/federation/ ). With this setup, Kiali could be configured to connect to this single federated Prometheus instance.

I wonder what would be your preference? Kiali connecting to a single federated prometheus? or Kiali connecting to multiple Prometheus to collect all needed data?

Also, if there is any reason why you can't use a federated Prometheus setup, I would like to hear your reasons to learn about people needs.

**Reply by ethernoy** [reply]
Hi @israel-hdez 

> I underestand that you have two meshes: one for the management cluster, and another one for cluster A and B. Do you really expect that Kiali shows a unified view of both meshes? Or do you expect Kiali to let you choose what mesh you want to inspect and then show only the selected mesh?

It looks more reasonable to me to choose the mesh to view since there should never be cross-mesh Istio traffic and hence it is meaningless to view all meshes in one graph.

> I wonder what would be your preference? Kiali connecting to a single federated prometheus? or Kiali connecting to multiple Prometheus to collect all needed data?

I would say both solutions on Prometheus (namely federation and multi-Prometheus) are OK for me (actually I was going to try a federated Istio Prometheus before you replied to me). What I concern more about is **whether it is necessary to let Kiali talk to Istiod from different clusters**. According to [Istio Multi-primary Multicluster Guideline](https://istio.io/latest/docs/setup/install/multicluster/multi-primary/), there could be multiple Istiod within one mesh in this configuration:

![image](https://user-images.githubusercontent.com/15870761/116637603-ea897e80-a996-11eb-93e7-965a645b1c6e.png)

In this case, if we are going to monitor a multi-cluster mesh with multiple Istiod deployments, is it a must to let Kiali connect to all of these Istiod pods?

**Reply by israel-hdez** [reply]
@ethernoy AFAIK, Kiali doesn't talk to Istiod pods.

The main dependencies are Prometheus, Jaeger (tracing) and the cluster API. Those are the deps we still need to connect to. It's still under discussion how to connect to these deps and still provide good performance.

**Comment by lucasponce** [comment]
El vie., 30 abr. 2021 19:36, Edgar Hernández ***@***.***>
escribió:

> @ethernoy <https://github.com/ethernoy> AFAIK, Kiali doesn't talk to
> Istiod pods.
>

That's not 100% correct.
Kiali calls istiod for several features:
- control plane status
- data plane proxy status

The main dependencies are Prometheus, Jaeger (tracing) and the cluster API.

**Reply by israel-hdez** [reply]
> That's not 100% correct. Kiali calls istiod for several features: - control plane status - data plane proxy status

Is it really connecting to `istiod`?

I thought control plane status was resolved querying the cluster API to get deployment status, and data plane proxy status I thought that was done via the sidecars rather than contacting `istiod`.

But if Kiali is directly connecting to `istiod`, then yes, it's a direct dependency.

**Reply by ethernoy** [reply]
I see. Then we need to resolve all these dependencies when Kiali is going to monitor multiple clusters.

I wonder if it is feasible to create a new page about direct dependency of data sources of Kiali on the documentation? It will be useful even for the current single cluster deployment.

**Comment by israel-hdez** [comment]
There is already one page in the website:

https://kiali.io/documentation/latest/architecture/

If I remember correctly, it was created even before Kiali v1.0 was released, quite a long time ago. It is still valid.

**Reply by ethernoy** [reply]
> There is already one page in the website: https://kiali.io/documentation/latest/architecture/ If I remember correctly, it was created even before Kiali v1.0 was released, quite a long time ago. It is still valid.

Oh thx! I did not find this page previously :0
